infra/imagestore/registry.go

```diff
@@ -50,93 +51,127 @@ type Registry struct {
+	var errors *multierror.Error
+	for _, auth := range i.registryAuths() {
+		manblob, err := imgcopy.Image(
+			ctx, i.polctx, toref, srcref, &imgcopy.Options{
+				ImageListSelection: imgcopy.CopyAllImages,
+				SourceCtx:          srcctx,
+				DestinationCtx: &types.SystemContext{
+					DockerInsecureSkipTLSVerify: insecure,
+					DockerAuthConfig:            auth,
+				},
+			},
+		)
+		if err != nil {
+			errors = multierror.Append(errors, err)
+			continue
+		}
+
+		dgst, err := manifest.Digest(manblob)
+		if err != nil {
+			return nil, fmt.Errorf("error calculating manifest digest: %w", err)
+		}
+
+		refstr := fmt.Sprintf("docker://%s@%s", toref.DockerReference().Name(), dgst)
+		return alltransports.ParseImageName(refstr)
 	}
 
-	refstr := fmt.Sprintf("docker://%s@%s", toref.DockerReference().Name(), dgst)
-	return alltransports.ParseImageName(refstr)
+	return nil, fmt.Errorf("unable to load image: %w", errors)
 }
 
-// Save pulls an image from our mirror registry, stores it in a temporary
-// tar file on disk.  Returns an ImageReference pointing to the local tar
-// file and a function the caller needs to call in order to clean up after
-// our mess (properly close tar file and delete it from disk). Returned ref
-// points to a 'docker-archive' tar file.
+// registryAuths returns a list of auths to be used when attempting to connect to the backend
+// registry. If no auth was configured for the backend registrythis function returns an slice
+// with a "nil" entry that, in containers/image/v5 library, means no auth (anonymous access).
+func (i *Registry) registryAuths() []*types.DockerAuthConfig {
+	if len(i.auths) == 0 {
+		return []*types.DockerAuthConfig{nil}
+	}
+	return i.auths
+}
+
+// Save pulls an image from our backend registry, stores it in a temporary tar file on disk.
+// Returns an ImageReference pointing to the local tar file and a function the caller needs to
+// call in order to clean up after our mess (properly close tar file and delete it from disk).
+// Returned ref points to a 'docker-archive' tar file.
 func (i *Registry) Save(
 	ctx context.Context, ref types.ImageReference,
 ) (types.ImageReference, CleanFn, error) {
 	domain := reference.Domain(ref.DockerReference())
 	if domain != i.regaddr {
-		return nil, nil, fmt.Errorf("mirror doesn't know about this image")
+		return nil, nil, fmt.Errorf("backend registry doesn't know about this image")
 	}
 
-	dstref, cleanup, err := i.NewLocalReference()
-	if err != nil {
-		return nil, nil, fmt.Errorf("error creating temp file: %w", err)
+	insecure := types.OptionalBoolFalse
+	if i.insecure {
+		insecure = types.OptionalBoolTrue
 	}
 
-	_, err = imgcopy.Image(
-		ctx, i.polctx, dstref, ref, &imgcopy.Options{
-			SourceCtx: i.regctx,
-		},
-	)
-	if err != nil {
-		cleanup()
-		return nil, nil, fmt.Errorf("unable to copy image: %w", err)
+	var err error
+	var cleanup CleanFn
+	var destref types.ImageReference
```

Comment:
These variables don't escape the `for` loop scope, so they could just be instantiated with `destref, cleanup, err := i.NewLocalReference` on line 150, right?

Reply:
nice catch, yes.